File: Algorithm/14Week(2024.04.24 ~ 2024.04.30)/Programmers/programmers_258712_이상현.py
def solution(friends, gifts):
    N = len(friends)

    gift_table = [[0] * N for _ in range(N)]
    gift_score_table = [[0] * 2 for _ in range(N)]
    dict_ = {friends[i]: i for i in range(N)}
    result = [0] * N

    for gift in gifts:
        give, receive = gift.split()
        gift_table[dict_[give]][dict_[receive]] += 1

    for row in range(N):
        gift_score_table[row][0] = sum(gift_table[row])
        gift_score_table[row][1] = sum(gift_table[i][row] for i in range(N))

    for i in range(N - 1):
        for j in range(i + 1, N):
            if gift_table[i][j] > gift_table[j][i]:
                result[i] += 1
            elif gift_table[i][j] < gift_table[j][i]:
                result[j] += 1
            else:
                if gift_score_table[i][0] - gift_score_table[i][1] > gift_score_table[j][0] - gift_score_table[j][1]:
                    result[i] += 1
                elif gift_score_table[i][0] - gift_score_table[i][1] < gift_score_table[j][0] - gift_score_table[j][1]:
                    result[j] += 1

    answer = max(result)
    return answer

File: Algorithm/14Week(2024.04.24 ~ 2024.04.30)/Programmers/programmers_258712_이상현.java
import java.util.HashMap;

class Solution {
    static StringBuilder sb = new StringBuilder();
    
    public int solution(String[] friends, String[] gifts) {
        int N = friends.length;
    	int[][] gift_table = new int[N][N];
    	int[][] gift_score_table = new int[N][3];
    	int[] result = new int[N];
    	HashMap<String, Integer> name_table = new HashMap<>();
    	
    	for (int i = 0; i < N; i++) {
    		name_table.put(friends[i], i);
    	}
    	
    	for (String gift: gifts) {
    		String[] temp = gift.split(" ");
    		gift_table[name_table.get(temp[0])][name_table.get(temp[1])]++;
    	}
    	
    	for (int i = 0; i < N; i++) {
    		for (int j = 0; j < N; j++) {
    			gift_score_table[i][0] += gift_table[i][j];
    			gift_score_table[i][1] += gift_table[j][i];
    		}
    		gift_score_table[i][2] = gift_score_table[i][0] - gift_score_table[i][1];
    	}
    	
    	for (int i = 0; i < N - 1; i++) {
    		for (int j = i + 1; j < N; j++) {
    			if (gift_table[i][j] > gift_table[j][i]) {
    				result[i]++;
    			} else if (gift_table[i][j] < gift_table[j][i]) {
    				result[j]++;
    			} else {
    				if (gift_score_table[i][2] > gift_score_table[j][2]) {
    					result[i]++;
    				} else if (gift_score_table[i][2] < gift_score_table[j][2]) {
    					result[j]++;
    				}
    			}
    		}
    	}
    	
    	int answer = 0;
    	for (int score: result) {
    		if (score > answer) {
    			answer = score;
    		}
    	}
        return answer;
    }
}